addons/bus/static/src/services/worker_service.js

```diff
@@ -0,0 +1,124 @@
+import { browser } from "@web/core/browser/browser";
+import { registry } from "@web/core/registry";
+import { Deferred } from "@web/core/utils/concurrency";
+import { session } from "@web/session";
+import { isIosApp } from "@web/core/browser/feature_detection";
+
+export const WORKER_STATE = Object.freeze({
+    UNINITIALIZED: "UNINITIALIZED",
+    INITIALIZING: "INITIALIZING",
+    INITIALIZED: "INITIALIZED",
+    FAILED: "FAILED",
+});
+
+export const workerService = {
+    dependencies: ["bus.parameters"],
+
+    start(env, { "bus.parameters": params }) {
+        let worker;
+        let isUsingSharedWorker = browser.SharedWorker && !isIosApp();
+        let state = WORKER_STATE.UNINITIALIZED;
+        let connectionInitializedDeferred = null;
+
+        function startWorker() {
+            if (!connectionInitializedDeferred) {
+                connectionInitializedDeferred = new Deferred();
+            }
+            state = WORKER_STATE.INITIALIZING;
+            let workerURL = `${params.serverURL}/bus/websocket_worker_bundle?v=${session.websocket_worker_version}`;
+            if (params.serverURL !== window.origin) {
+                // Worker service can be loaded from a different origin than the
+                // bundle URL. The Worker expects an URL from this origin, give
+                // it a base64 URL that will then load the bundle via "importScripts"
+                // which allows cross origin.
+                const source = `importScripts("${workerURL}");`;
+                workerURL = "data:application/javascript;base64," + window.btoa(source);
+            }
+            const workerClass = isUsingSharedWorker ? browser.SharedWorker : browser.Worker;
+            worker = new workerClass(workerURL, {
+                name: isUsingSharedWorker ? "odoo:bus_shared_worker" : "odoo:bus_worker",
+            });
+            worker.onerror = onInitError;
+            _registerHandler((ev) => {
+                if (ev.data.type === "BASE:INITIALIZED") {
+                    state = WORKER_STATE.INITIALIZED;
+                    connectionInitializedDeferred.resolve();
+                }
+            });
+            if (isUsingSharedWorker) {
+                worker.port.start();
+            }
+            _send("BASE:INIT");
+        }
+
+        function onInitError() {
+            if (state === WORKER_STATE.INITIALIZING && isUsingSharedWorker) {
+                console.warn("Error while loading SharedWorker, fallback on Worker.");
+                isUsingSharedWorker = false;
+                worker?.port?.close?.();
+                startWorker();
+            } else if (state === WORKER_STATE.INITIALIZING) {
```

Comment:
That's weird for browser to support shared worker but it failed to initialize.
From my understanding we had it happen on some machines for unknown reasons.
Ok to keep this branching, but would put a `// FIXME` comment to clarify this is strange and something to investigate eventually.